File: lecture-notes/games-dev-pathway/week-02-design-patterns/code-examples/builder.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <random>
#include <string>
#include <stdexcept>
#include <format>

struct Card
{
    std::string suit;
    int value;
    std::string toString() const
    {
        std::string name;
        if (value == 11)
            name = "Jack";
        else if (value == 12)
            name = "Queen";
        else if (value == 13)
            name = "King";
        else if (value == 14)
            name = "Ace";
        else
            name = std::to_string(value);
        return name + " of " + suit;
    }
};

struct GameConfig
{
    std::string player1Name = "Player 1";
    std::string player2Name = "Player 2";
    int maxRounds = 26;
    bool shuffleDeck = true;
    bool tieGoesToP1 = false;
    bool verbose = true;
};

class GameConfigBuilder
{
public:
    GameConfigBuilder &setPlayer1(const std::string &name)
    {
        config.player1Name = name;
        return *this;
    }

    GameConfigBuilder &setPlayer2(const std::string &name)
    {
        config.player2Name = name;
        return *this;
    }

    GameConfigBuilder &setMaxRounds(int rounds)
    {
        if (rounds < 1 || rounds > 26)
            throw std::invalid_argument("Rounds must be between 1 and 26");
        config.maxRounds = rounds;
        return *this;
    }

    GameConfigBuilder &withShuffle(bool shuffle)
    {
        config.shuffleDeck = shuffle;
        return *this;
    }

    GameConfigBuilder &withTieRuleP1()
    {
        config.tieGoesToP1 = true;
        return *this;
    }

    GameConfigBuilder &withVerboseOutput(bool v)
    {
        config.verbose = v;
        return *this;
    }

    GameConfig build() const
    {
        return config;
    }

private:
    GameConfig config;
};

class BuilderWarGame
{
public:
    void play(const GameConfig &cfg)
    {
        if (cfg.verbose)
            std::cout << std::format("Starting game: {} vs {}  ({} rounds, shuffle={})\n\n",
                                     cfg.player1Name, cfg.player2Name, cfg.maxRounds, cfg.shuffleDeck);

        std::vector<Card> deck;
        for (const auto &suit : {"Spades", "Hearts", "Clubs", "Diamonds"})
            for (int v = 2; v <= 14; ++v)
                deck.push_back({suit, v});

        if (cfg.shuffleDeck)
        {
            std::mt19937 rng(std::random_device{}());
            std::shuffle(deck.begin(), deck.end(), rng);
        }

        std::vector<Card> hand1(deck.begin(), deck.begin() + 26);
        std::vector<Card> hand2(deck.begin() + 26, deck.end());

        int w1 = 0, w2 = 0;
        for (int round = 1; round <= cfg.maxRounds; ++round)
        {
            Card c1 = hand1.back();
            hand1.pop_back();
            Card c2 = hand2.back();
            hand2.pop_back();

            std::string outcome;
            if (c1.value > c2.value)
            {
                ++w1;
                outcome = cfg.player1Name + " wins";
            }
            else if (c2.value > c1.value)
            {
                ++w2;
                outcome = cfg.player2Name + " wins";
            }
            else
            {

                if (cfg.tieGoesToP1)
                {
                    ++w1;
                    outcome = "Tie => " + cfg.player1Name + " (house rule)";
                }
                else
                {
                    outcome = "Tie";
                }
            }

            if (cfg.verbose)
                std::cout << std::format("Round {:2}: {} vs {}  =>  {}\n",
                                         round, c1.toString(), c2.toString(), outcome);
        }

        std::cout << std::format("\nFinal: {} = {}  {} = {}\n",
                                 cfg.player1Name, w1, cfg.player2Name, w2);
    }
};

int main()
{

    GameConfig cfg = GameConfigBuilder()
                         .setPlayer1("Alice")
                         .setPlayer2("Bob")
                         .setMaxRounds(10)
                         .withShuffle(true)
                         .withTieRuleP1()
                         .withVerboseOutput(true)
                         .build();

    BuilderWarGame game;
    game.play(cfg);

    return 0;
}

File: lecture-notes/app-dev-pathway/week-02-design-patterns/code-examples/builder.py
import random


class Card:
    def __init__(self, suit, value):
        self.suit = suit
        self.value = value

    def __str__(self):
        names = {11: "Jack", 12: "Queen", 13: "King", 14: "Ace"}
        name = names.get(self.value, str(self.value))
        return f"{name} of {self.suit}"


class GameConfig:
    def __init__(self):
        self.player1_name = "Player 1"
        self.player2_name = "Player 2"
        self.max_rounds = 26
        self.shuffle_deck = True
        self.tie_goes_to_p1 = False
        self.verbose = True


class GameConfigBuilder:
    def __init__(self):
        self.config = GameConfig()

    def set_player1(self, name):
        self.config.player1_name = name
        return self

    def set_player2(self, name):
        self.config.player2_name = name
        return self

    def set_max_rounds(self, rounds):
        if rounds < 1 or rounds > 26:
            raise ValueError("Rounds must be between 1 and 26")
        self.config.max_rounds = rounds
        return self

    def with_shuffle(self, shuffle):
        self.config.shuffle_deck = shuffle
        return self

    def with_tie_rule_p1(self):
        self.config.tie_goes_to_p1 = True
        return self

    def with_verbose_output(self, verbose):
        self.config.verbose = verbose
        return self

    def build(self):
        return self.config


class BuilderWarGame:
    def play(self, cfg):
        if cfg.verbose:
            print(f"Starting game: {cfg.player1_name} vs {cfg.player2_name}  "
                  f"({cfg.max_rounds} rounds, shuffle={cfg.shuffle_deck})\n")

        suits = ["Spades", "Hearts", "Clubs", "Diamonds"]
        deck = [Card(suit, value) for suit in suits for value in range(2, 15)]

        if cfg.shuffle_deck:
            random.shuffle(deck)

        hand1 = deck[:26]
        hand2 = deck[26:]

        w1 = 0
        w2 = 0

        for round_num in range(1, cfg.max_rounds + 1):
            c1 = hand1.pop()
            c2 = hand2.pop()

            if c1.value > c2.value:
                w1 += 1
                outcome = f"{cfg.player1_name} wins"
            elif c2.value > c1.value:
                w2 += 1
                outcome = f"{cfg.player2_name} wins"
            else:
                if cfg.tie_goes_to_p1:
                    w1 += 1
                    outcome = f"Tie => {cfg.player1_name} (house rule)"
                else:
                    outcome = "Tie"

            if cfg.verbose:
                print(f"Round {round_num:2}: {c1} vs {c2}  =>  {outcome}")

        print(f"\nFinal: {cfg.player1_name} = {w1}  {cfg.player2_name} = {w2}")


cfg = (GameConfigBuilder()
       .set_player1("Alice")
       .set_player2("Bob")
       .set_max_rounds(10)
       .with_shuffle(True)
       .with_tie_rule_p1()
       .with_verbose_output(True)
       .build())

game = BuilderWarGame()
game.play(cfg)